main.tf

```diff
@@ -314,52 +665,334 @@ resource "azurerm_orchestrated_virtual_machine_scale_set" "virtual_machine_scale
-  notes      = var.lock.kind == "CanNotDelete" ? "Cannot delete the resource or its child resources." : "Cannot delete or modify the resource or its child resources."
+  name           = module.avm_utl_interfaces.lock_azapi.name != null ? module.avm_utl_interfaces.lock_azapi.name : "lock-${azapi_resource.virtual_machine_scale_set.name}"
+  parent_id      = azapi_resource.virtual_machine_scale_set.id
+  type           = module.avm_utl_interfaces.lock_azapi.type
+  body           = module.avm_utl_interfaces.lock_azapi.body
+  create_headers = var.enable_telemetry ? { "User-Agent" : local.avm_azapi_header } : null
+  delete_headers = var.enable_telemetry ? { "User-Agent" : local.avm_azapi_header } : null
+  read_headers   = var.enable_telemetry ? { "User-Agent" : local.avm_azapi_header } : null
+  update_headers = var.enable_telemetry ? { "User-Agent" : local.avm_azapi_header } : null
+
+  depends_on = [azapi_resource.role_assignments]
+}
+
+moved {
+  from = azurerm_role_assignment.this
+  to   = azapi_resource.role_assignments
+}
+
+resource "azapi_resource" "role_assignments" {
+  for_each = var.role_assignments
+
+  name      = module.avm_utl_interfaces.role_assignments_azapi[each.key].name
+  parent_id = azapi_resource.virtual_machine_scale_set.id
+  type      = module.avm_utl_interfaces.role_assignments_azapi[each.key].type
+  body = {
+    properties = {
+      principalId                        = module.avm_utl_interfaces.role_assignments_azapi[each.key].body.properties.principalId
+      roleDefinitionId                   = module.avm_utl_interfaces.role_assignments_azapi[each.key].body.properties.roleDefinitionId
+      condition                          = module.avm_utl_interfaces.role_assignments_azapi[each.key].body.properties.condition
+      conditionVersion                   = module.avm_utl_interfaces.role_assignments_azapi[each.key].body.properties.conditionVersion
+      delegatedManagedIdentityResourceId = module.avm_utl_interfaces.role_assignments_azapi[each.key].body.properties.delegatedManagedIdentityResourceId
+      description                        = module.avm_utl_interfaces.role_assignments_azapi[each.key].body.properties.description == null ? "" : module.avm_utl_interfaces.role_assignments_azapi[each.key].body.properties.description
+      principalType                      = module.avm_utl_interfaces.role_assignments_azapi[each.key].body.properties.principalType
+    }
+  }
+  create_headers = var.enable_telemetry ? { "User-Agent" : local.avm_azapi_header } : null
+  delete_headers = var.enable_telemetry ? { "User-Agent" : local.avm_azapi_header } : null
+  read_headers   = var.enable_telemetry ? { "User-Agent" : local.avm_azapi_header } : null
+  retry = {
+    error_message_regex = [
+      ".*Please remove the lock and try again.*",
+    ]
+  }
+  update_headers = var.enable_telemetry ? { "User-Agent" : local.avm_azapi_header } : null
+
+  lifecycle {
```

Comment:
please remove this, and the update resource and set ignore_null_property=true on this resource to mimic the behaviour of azurerm's optional_computed.